### PR TITLE
Final version of white-paper

### PR DESCRIPTION
**Pull request by kentr0w**
### What's done:
    Final version of white paper

### CODE REVIEW

**Review by petertrr — CHANGES_REQUESTED** [review]
Please also add build instructions for latex, either with `Makefile` like in other cqfn projects, or with gradle task (for now in the same project in `info` directory). If you go with make, you can add a gradle task that calls it as well.

File: wp/sections/work.tex [hunk 1]
```diff
@@ -188,7 +207,55 @@ \subsection{Examples of unique inspections}
...
 \end{tikzpicture}
-    \item \textbf{FLOAT\underline{ }IN\underline{ }ACCURATE\underline{ }CALCULATIONS} - Rule that checks that floating-point numbers are not used for accurate calculations.
-\end{enumerate}
+\end{center}
+    
+In order to collect  elements with a proper type, Inspection has to do a tree traversal. Tree traversal is done by a special method called \texttt{findAllNodesWithCondition()} (see Listing 3). This function searches for a node with a given condition (in this case it is when node's type equals to \texttt{IDENTIFIER}). As a basis it uses DFS (Depth-first search): it goes recursively in depth of the tree and compares types of AST nodes. When it finds necessary node it returns it and the result from it's parent search as a list, otherwise it returns empty list. At the end the Inspection checks the package name based on identifiers and file name. 
+
+
+\begin{lstlisting}[caption={Method for AST traversal}, label={lst:example1}, language=Kotlin]      
+**
```
Comment (petertrr) [inline wp/sections/work.tex:249]:
```suggestion
/**
```

File: wp/sections/feature.tex [hunk 2]
```diff
@@ -95,13 +96,13 @@ \subsection{"SUPPRESS annotation"}
...
-There will be warning: 
+Diktat will raise the warning: 
 $$
 \texttt{ \small{ $\big[$FUNCTION\underline{ }NAME\underline{ }INCORRECT\underline{ }CASE$\big]$  function/method name should be in lowerCamelCase}}
 $$
 
-But if there is a \textsl{@SUPPRESS} before this method, then there will be no errors at run.
-\begin{lstlisting}[caption={Function with incorrect name, but with suppress.}, label={lst:example1}, language=Kotlin]
+But if there is a \texttt{@Suppress} before this method, then there will be no warnings during the run:
+\begin{lstlisting}[caption={Function with incorrect name, but with suppressed Inspection.}, label={lst:example1}, language=Kotlin]
```
Comment (petertrr) [inline wp/sections/feature.tex:105]: Please align asterisks in KDocs properly in this listing.

File: wp/sections/feature.tex [hunk 3]
```diff
@@ -1,4 +1,4 @@
-\par As described above, diKTat is very configurable and user-friendly. But these are not all of its advantages and features. Below will be presented and described unusual and important killer-features of diKTat.
+\par As described above, diKTat is configurable and user-friendly. But these are not all of it's advantages and features. Below we will present and describe unusual and important killer-features of diKTat.
 
 \subsection{Configuration file}
 \par
```
Comment (petertrr) [inline wp/sections/feature.tex:4]: Is there syntax highlighting for yml in latex? You now have `language=Kotlin` it this block.

File: wp/sections/download.tex [hunk 4]
```diff
@@ -1,8 +1,76 @@
...
+                        </goals>
+                        <configuration>
+                            <inputs>
+                                <input>${project.basedir}/src/main/kotlin</input>
+                                <input>${project.basedir}/src/test/kotlin</input>
+                            </inputs>
+                            <diktatConfigFile>diktat-analysis.yml</diktatConfigFile>
+                           <excludes>
+                              <exclude>${project.basedir}/src/test/kotlin/excluded</exclude>
+                           </excludes>
```
Comment (petertrr) [inline wp/sections/download.tex:39]:
```suggestion
                            <excludes>
                                <exclude>${project.basedir}/src/test/kotlin/excluded</exclude>
                            </excludes>
```

File: wp/sections/download.tex [hunk 5]
```diff
@@ -1,8 +1,76 @@
...
+
+\subsection{Gradle plugin}
+This plugin is available since version 0.1.5. The following code snippet shows how to configure Gradle plugin for diktat.
+
+\begin{lstlisting}[caption={DiKTat with Gradle plugin}, label={lst:gradle1}, language=Kotlin]
+plugins {
+    id("org.cqfn.diktat.diktat-gradle-plugin") version "0.1.7"
+}
+
+apply(plugin = "org.cqfn.diktat.diktat-gradle-plugin")
```
Comment (petertrr) [inline wp/sections/download.tex:57]:
```suggestion
```
This line is not needed if using plugin DSL, it's from the second option with buildscript.

**Review by petertrr — CHANGES_REQUESTED** [review]

File: wp/sections/appendix.tex [hunk 6]
```diff
@@ -151,9 +147,9 @@ \subsubsection*{\textbf{1.1.1 Identifiers naming conventions}}
 
 For identifiers, use the following naming conventions:
 
-1.	All identifiers should use only ASCII letters or digits, and the names should match regular expressions `\w{2,64}`.
+1.	All identifiers should use only ASCII letters or digits, and the names should match regular expressions `{2,64}`.
```
Comment (petertrr) [inline wp/sections/appendix.tex:150]: What's with this `\w`?

**Review by petertrr — CHANGES_REQUESTED** [review]
There are trailing semicolons in examples in code style guide, in markdown too. Especially in the first chapter. They should be removed.

**Comment by petertrr** [comment]
Links (in form [text](link)) are not converted from markdown into tex.

**Comment by kentr0w** [comment]
Bugs with semicolon and link will be corrected in another PR

File: README.md [hunk 7]
```diff
@@ -201,3 +201,13 @@ $ mvn clean install
 ```
 
 Also see our [Contributing Policy](CONTRIBUTING.md) and [Code of Conduct](CODE_OF_CONDUCT.md)
+
+## White paper
```
Comment (petertrr) [inline README.md:205]: I think we can have a separate readme inside `wp` directory with these build instructions, not to overload the main readme.

File: wp/makefile [hunk 8]
```diff
@@ -0,0 +1,5 @@
+all:
+	xelatex wp.tex
```
Comment (petertrr) [inline wp/makefile:2]: How did you test this command? I tried to run `make all` on Ubuntu, and it entered interactive mode with a bunch of errors.

**Review by petertrr — CHANGES_REQUESTED** [review]
Let's not upload wp.pdf to git, but instead add a link like `https://github.com/cqfn/diKTat/releases/download/wp/wp.pdf` to the main README.

File: wp/makefile (hunk 8 above)
Comment (petertrr) [reply wp/makefile:2]: Maybe [this](https://www.overleaf.com/learn/latex/chinese#CJKutf8_with_pdfLaTeX) could help?
> this is usually only convenient for documents in English with bits of Chinese text

Comment (kentr0w) [reply wp/makefile]: Yes, thank you!

File: info/buildSrc/src/main/kotlin/org/cqfn/diktat/generation/docs/GenerationDocs.kt [hunk 9]
```diff
@@ -5,6 +5,8 @@ package org.cqfn.diktat.generation.docs
 import java.io.File
 import java.io.PrintWriter
 
+
+var qwe = 0
```
Comment (petertrr) [inline info/buildSrc/src/main/kotlin/org/cqfn/diktat/generation/docs/GenerationDocs.kt:9]: ?

**Review by petertrr — COMMENTED** [review]
Three things to do and this PR should be merged:
1. Build and upload pdf to wp directory (let's keep things simple for now)
2. Add note about miktex version
3. Fix table in root README with link to WP

**Review by petertrr — APPROVED** [review]

**Pull request merged**